### PR TITLE
fancysets: fix for special sets (Integers, Naturals)

### PR DESCRIPTION
**Pull request by rohitrango**
Returns S.false for comparing intervals and sets against these sets
`S.Integers`, `S.Naturals`, `S.Naturals0`, `S.Complexes`
Instead of throwing back errors

Example:

```
In [1]: Interval(1,2) in S.EmptySet
Out[1]: False

In [2]: Interval(1,2) in S.Integers
Out[2]: False

In [3]: Interval(1,2) in S.Naturals
Out[3]: False

In [4]: Interval(1,2) in S.Naturals0
Out[4]: False

In [5]: EmptySet() in S.Naturals0
Out[5]: False

In [6]: EmptySet() in S.Naturals
Out[6]: False

In [7]: Interval(1,2) in S.Reals
Out[7]: False

In [8]: Interval(1,2) in S.Complexes
Out[8]: False
```

### CODE REVIEW

**Comment by rohitrango** [comment]
#11732

**Comment by jksuom** [comment]
We could exclude all sets with a single `if isinstance(other, Set)` instead of listing different types of sets, but probably even better would be to exclude all non-numbers using `if not other.is_number` before other tests.

**Comment by rohitrango** [comment]
Yes, initially I had those checks as I was considering the idea of using it as a subset. 
Can you please tell why are there so many failed checks?

**Comment by rohitrango** [comment]
Im getting these errors upon running tests for `./bin/test sets`

```
________________________________________________________________________________
____________ sympy/sets/tests/test_contains.py:test_contains_basic _____________
  File "/home/rohitrango/Desktop/scipy/sympy/sympy/sets/tests/test_contains.py", line 9, in test_contains_basic
    assert Contains(i, S.Naturals) == Contains(i, S.Naturals, evaluate=False)
AssertionError
________________________________________________________________________________
______________ sympy/sets/tests/test_contains.py:test_issue_6194 _______________
  File "/home/rohitrango/Desktop/scipy/sympy/sympy/sets/tests/test_contains.py", line 17, in test_issue_6194
    assert Contains(x, S.Integers) != S.false
AssertionError
```

For the second one, i think I should not evaluate for symbols and just return the unevaluated `Contains` form of it. Am I right?

**Comment by rohitrango** [comment]
@certik @asmeurer Can you tell me where am I going wrong? A little guidance would be really helpful.

**Comment by asmeurer** [comment]
Yes, symbol objects (an expressions in general) should remain unevaluated in Contains.

**Comment by asmeurer** [comment]
I would use @jksuom's suggestion.

**Comment by rohitrango** [comment]
Could you please review it now? I only had to add an elif-case for the `Set` class.
But the only thing I'm unable to do right now is to add this feature (by adding a `_contains` method for `Complexes`) for S.Complexes class (it breaks some tests). I'm unaware of all the assumptions that are related to this class. Could you tell me what things I need to keep in mind while dealing with the `S.Complex` class?

**Comment by rohitrango** [comment]
@jksuom Please review it and suggest changes, if any.

**Comment by jksuom** [comment]
I think it would be better to change the order of the tests because `.is_integer` is based on the assumption system and may be time consuming. There is no need to apply that test to objects that cannot possibly be integers, such as sets. Hence testing for such objects should be done before `.is_integer`.

Another matter is that there are other non-numeric objects in addition to sets. It would be advisable to return `False` for all of them. That could probably be done by something like `if other.is_number is False`.

**Comment by rohitrango** [comment]
That makes it a little risky actually. I tried it before, and it failed a lot of tests. Symbols and expressions have to be handled separately, and I'm not completely sure what other types of objects it must handle and what type of objects it should ignore.
For example,
```
In [2]: x.is_number
Out[2]: False
```

**Comment by jksuom** [comment]
Perhaps  expressions could be dealt with by adding `not isinstance(other, Expr)`, which also a fairly cheap test.

**Comment by rohitrango** [comment]
I do not think this would work. I have tried many things and the code seems to be breaking upon trying to check for different types of SymPy objects. I think it is best to implement this to `Sets` for now.

**Comment by jksuom** [comment]
>  the code seems to be breaking upon trying to check for different types of SymPy objects

Can you explain in more detail how the code is breaking and for which kind on SymPy objects that will happen?

**Comment by rohitrango** [comment]
It seems that many objects have `Expr` as their base class, and the problem occurs mainly when I modify the `ComplexRegion` class. It fails 14 cases, none of which are from the module `Sets`. This is the code that I had added in the `ComplexRegion` class:

```
if isinstance(other, Expr) and other.is_number is False:
       return None
if isinstance(other,Set):
        return False
```

It fails 14 tests, one of them is the `./bin/test core/tests/test_function`, which gives this error- 
```
________________________________________________________________________________
______________ sympy/core/tests/test_function.py:test_issue_7231 _______________
  File "/home/rohitrango/Desktop/scipy/sympy/sympy/core/tests/test_function.py", line 656, in test_issue_7231
    ans1 = f(x).series(x, a)
  File "sympy/core/expr.py", line 2506, in series
    return s.subs(x, rep2 + rep2b)
  File "sympy/core/basic.py", line 910, in subs
    rv = rv._subs(old, new, **kwargs)
  File "sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 807, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/core/basic.py", line 1024, in _subs
    rv = fallback(self, old, new)
  File "sympy/core/basic.py", line 996, in fallback
    arg = arg._subs(old, new, **hints)
  File "sympy/core/cache.py", line 93, in wrapper
    retval = cfunc(*args, **kwargs)
  File "sympy/core/compatibility.py", line 807, in wrapper
    result = user_function(*args, **kwds)
  File "sympy/core/basic.py", line 1022, in _subs
    rv = self._eval_subs(old, new)
  File "sympy/series/order.py", line 446, in _eval_subs
    point = d.subs(res[0]).limit(old, self.point[i])
AttributeError: 'Contains' object has no attribute 'limit'

= tests finished: 51 passed, 2 expected to fail, 1 exceptions, in 2.96 seconds =
DO *NOT* COMMIT!
```
I do not see how the `ComplexRegion` class affects this. I think its because of some other class being used and which is a subclass of `Expr`. If possible, could you please give me a hierarchy of the SymPy objects and a place to start with the current issue?

**Comment by jksuom** [comment]
```
if isinstance(other, Expr) and other.is_number is False:
    return None
```
What I had in mind was something along these lines:
```
if not isinstance(other, Expr):
    return False
```
I assume that all numerical objects in SymPy belong to (a subclass of) Expr. But if that is not the case, then some additional test (presumably involving `.is_number`) would be needed.


It seems to me that the hierarchy of SymPy objects is too complicated to describe. Besides, it is not stable but new classes are added and some old ones are removed all the time.

If I want to see where specific class is located in the hierarchy, I use the `__mro__` attribute of the class.
```
In [2]: Interval.__mro__
Out[2]: 
(sympy.sets.sets.Interval,
 sympy.sets.sets.Set,
 sympy.core.basic.Basic,
 sympy.core.evalf.EvalfMixin,
 object)
```

**Comment by rohitrango** [comment]
The thing is even `Integer(1)` and `Rational(1/2)` are instances of type `Expr`. So a check other than `Expr` would be required.  
So, what should I be working on now?

**Comment by jksuom** [comment]
> The thing is even Integer(1) and Rational(1/2) are instances of type Expr. So a check other than Expr would be required.

We want a quick test that would recognize objects that cannot possibly be numbers. Sets are such objects, for example. But they are not the only ones. It seems that all objects that can denote numbers are expressions. So, if we want to test whether `other` is an integer or not, we can start with the following piece of code:
```
if not isinstance(other, Expr):
    return False
```
instead of
```
if not isinstance(other, Set):
    return False
```
The latter test will only work for sets, but the former one is more general. It will potentially cover even such classes of non-numerical objects that have not yet been defined.

After this preliminary test only expressions such as `Integer(1)` and `Rational(1, 2)` remain to be further investigated.

**Comment by rohitrango** [comment]
Right! I was thinking about it in a more complicated way. Thanks for the clarification. I have made the changes, and I'm testing it on my own system before pushing the changes. The build process takes a lot of time though :confused:

**Comment by jksuom** [comment]
This looks good apart from minor PEP 8 issues.

File: sympy/sets/fancysets.py [hunk 1]
```diff
@@ -54,7 +54,9 @@ def _intersect(self, other):
         return None
 
     def _contains(self, other):
-        if other.is_positive and other.is_integer:
+        if not isinstance(other,Expr):
```
Comment (rohitrango) [inline sympy/sets/fancysets.py]: Ohh right!

**Comment by jksuom** [comment]
The code looks good, but we should also have some tests referring to #11732. (`test_issue_11732()` in sympy/sets/tests/test_fancysets.py.)

**Comment by rohitrango** [comment]
Sure! Could you guide me a bit on writing tests, considering I haven't written one before?

**Comment by jksuom** [comment]
>  Do I only have to write assert statements with the correct assertions?

Yes, that is the general idea. You should write tests that show how the behaviour has been improved.
(No more errors raised.)

**Comment by rohitrango** [comment]
@jksuom I think these tests will be good.

**Comment by jksuom** [comment]
Thanks. This is ready to be merged.

**Pull request merged**